src/geometry/manifold/basic_smooth_bundle.lean

```diff
@@ -0,0 +1,568 @@
+coordinate charts. It can be constructed from a basic smooth bundled core, defined below, specifying
+the changes in the fiber when one goes from one coordinate chart to another one.
+
+## Main definitions
+
+`basic_smooth_bundle_core I M F`: M is a smooth manifold over the model with corners I on (𝕜, E, H),
+       and F is a normed vector space over 𝕜. This structure registers, for each pair of charts
+       of M, a smooth change of coordinates on F. This is the core structure from which one will
+       build a smooth vector bundle with fiber F over M.
+Let `Z` be a basic smooth bundle core over M with fiber F. We define
+`Z.to_topological_fiber_bundle_core`, the (topological) fiber bundle core associated to Z. From it,
+we get a space `Z.to_topological_fiber_bundle_core.total_space` (which as a Type is just M × F),
+with the fiber bundle topology. It inherits a manifold structure (where the charts are in bijection
+with the charts of the basis). We show that this manifold is smooth.
+
+Then we use this machinery to construct the tangent bundle of a smooth manifold.
+
+* `tangent_bundle_core I M`: the basic smooth bundle core associated to a smooth manifold M over a
+                             model with corners I.
+* `tangent_bundle I M`     : the total space of `tangent_bundle_core I M`. It is itself a
+                             smooth manifold over the model with corners `I.tangent`, the product of
+                             `I` and the trivial model with corners on `E`.
+* `tangent_space I x`      : the tangent space to M at x
+* `tangent_bundle.proj I M`: the projection from the tangent bundle to the base manifold
+
+## TODO
+construct the cotangent bundle, and the bundles of differential forms. They should follow
+functorially from the description of the tangent bundle as a basic smooth bundle.
+
+## Tags
+Smooth fiber bundle, vector bundle, tangent space, tangent bundle
+-/
+
+noncomputable theory
+
+universe u
+
+open topological_space set
+
+/-- Core structure used to create a smooth bundle above M (a manifold over the model with
+corner I) with fiber the normed vector space F over 𝕜, which is trivial in the chart domains of M.
+This structure registers the changes in the fibers when one changes coordinate charts in the base. -/
+structure basic_smooth_bundle_core {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type u} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(M : Type*) [topological_space M] [manifold H M] [smooth_manifold_with_corners I M]
+(F : Type u) [normed_group F] [normed_space 𝕜 F] :=
+(coord_change      : atlas H M → atlas H M → H → F → F)
+(coord_change_self :
+  ∀i : atlas H M, ∀ x ∈ i.1.target, ∀v, coord_change i i x v = v)
+(coord_change_comp : ∀i j 𝕜 : atlas H M,
+  ∀x ∈ ((i.1.symm.trans j.1).trans (j.1.symm.trans 𝕜.1)).source, ∀v,
+  (coord_change j 𝕜 ((i.1.symm.trans j.1).to_fun x)) (coord_change i j x v) = coord_change i 𝕜 x v)
```

Comment:
Did you think about introducing some cocycle definition to make that more readable?

Reply:
The issue is that the cocycle relation should only hold on the right domain, so getting a nice usable general definition is not that obvious. Or I could copy exactly the definition here, but I don't think it would buy me much.